### PR TITLE
Feat:  CLI Enhancements, Multi-GPU Inference

### LINKED ISSUE #95: [Feature] Allow passing file to inference on
**Problem**

It may be necessary to repeat the same questions across many experiments. It is time consuming to copy paste line by line.

**Feature**

Allow passing path to jsonl file or similar that can be read and ran through the model then output to a results file

### LINKED ISSUE #165: [Fix] Deprecate selecting yaml from list
A user is getting this error when using this feature with MULTI-GPU.

https://discord.com/channels/1104757954588196865/1111279858136383509/1116379061875376138

Since the issue comes from passing and not passing the config during CLI, I suspect it's from this feature where only one GPU will be set and others will be waiting.

https://github.com/OpenAccess-AI-Collective/axolotl/blob/81911d112c773739a721032f92f14b53d5f5d852/scripts/finetune.py#L108-L131

### LINKED ISSUE #292: Discuss: Axolotl CLI, sub-modules, and configuration
Good morning, I want to start a discussion with the community regarding the CLI and how actions can be launched. Axolotl currently supports several actions including: fine-tuning, dataset prep, merge lora, and inferencing. Over time, as more features are added, the logic currently in finetune.py / train could get more complicated as new actions are added. To start a discussion, here are a few thoughts in no particular order:

* Configuration defaults & overrides for example: configuration file, environment variables, CLI options. IMO a consistent hierarchy of configuration defaults & overrides will make it easier for folks to run Axolotl in multiple steps in more complicated pipelines.

* Increase pytest testability: Any refactoring of the CLI should promote as much modularity as possible so individual components can be tested in isolation

* Possible decision: Do we want 1 entrypoint (ex ``finetune.py``) or multiple entrypoints: (ex: ``batch_eval.py``, ``batch_inference.py``, etc.)

* I checked and Fire supports sub-commands and basic help generation, ex:

```python
if __name__ == "__main__":

    fire_config = {
        "train": {
            "function": train,
            "args": {
                "config": {
                    "help": "Path to the configuration file",
                },
            },
            "description": "Invoke finetuning",
        },
        "evaluate": {
            "function": evaluate,
            "args": {...},
            "description": "Perform batch evaluation",
        },
        "inference": {
            "function": inference,
            "args": {...},
            "description": "Interactively perform inferencing",
        },
        "batch_eval": {
            "function": batch_eval,
            "args": {...},
            "description": "Interactively perform inferencing",
        },
    }

    import sys
    if len(sys.argv) >= 2 and sys.argv[1] in ["--help", *fire_config.keys()]:
        # when the first parameter is a command group or "--help" invoke the command group CLI parser
        fire.Fire(fire_config)
    else:
        # Else, invoke backward-compatibility CLI behavior
        fire.Fire(train)
```

* While I personally prefer a single entrypoint with multiple sub-commands, I believe there is a very strong argument to be made that the greater goal is to make it as easy as possible for folks to implement bleeding-edge capabilities quickly. Too much structure in the wrong places could slow things down.

* If there is an appetite to go with multiple sub-commands, maybe [Click](https://click.palletsprojects.com/en/8.1.x/) would be a better option. While I'm currently learning Fire, Click does have good support for subcommands, generating CLI help documentation, validation, CLI option types, and integrating environment variable overrides.

Thanks to the Axolotl community for your hard work and great feedback!

### PR DESCRIPTION
**Pull request by theobjectivedad**
This update adds the following changes:

* New feature: Axolotl CLI enhancements
* New feature:  Multi-GPU evaluation
* fixes #95 
* fixes #292
* fixes #165 

## Axolotl CLI enhancements

Today, Axolotl offers a wide range of functionalities that facilitate the finetuning of Large Language Models (LLMs). These include dataset preparation, creating and saving a merged (q)LoRA model, data sharding, interactive inferencing, and the finetuning process itself.

Looking ahead, a typical, albeit not comprehensive, lifecycle sequence for LLM finetuning might be:

```text
Dataset Preparation -> Finetuning -> Quantization & Export -> Validation & Benchmarking -> Inferencing
```

The CLI update is intended to significantly simplify the realization of such lifecycle chains, for both end users and developers:

* **Configuration Overrides**: In accordance with the [12-factor app best practices](https://12factor.net/config) best practices, configuration values can now be overridden by CLI parameters and environment variables. This adds a standard hierarchy of overrides where defaults reside in the current ``config.yaml`` file, which can be overridden by environment variables, which can be overridden by CLI parameters.
* **Easier for end-users**: Considering that not all configuration options are pertinent to all features—for instance, "evaluation" only necessitates a subset of "finetune" settings—this update establishes a framework that makes it easier for users to identify the settings used by each feature. Moreover, it provides online assistance for each setting (accessible via CLI --help). Additionally, the update ensures full backward compatibility with ``finetune.py``, and capitalizes on existing ``config.yaml`` settings are already integrated into Axolotl.
* **Easier for contributors**: The CLI enhancements provide a clear way for contributors to add new and modular pipeline steps 
* **Orchestration**: An overridable configuration hierarchy will lay the foundation for a more automation of Axolotl pipeline steps via external orchestration tools, such as [Apache Airflow](https://airflow.apache.org/).

The CLI enhancements implement both evaluation and inferencing pipeline steps as useful examples. The core finetuning step remain in ``finetune.py`` and are fully compatible with this update. Moreover, all existing Axolotl configuration options are intact and are considered foundational to the new CLI. As needed, the finetuning can be moved into the new CLI in a future PR.

### Example CLI outputs:

Main CLI help:

```shell
Usage: axolotl [OPTIONS] COMMAND [ARGS]...

  Axolotl CLI

Options:
  --config/-c PATH                   Path to configuration file, if set to a
                                  directory axolotl will prompt for the config
                                  file  [env var: AXOLOTL_CONFIG; default:
                                  configs]
  --log_level [DEBUG|INFO|WARNING|ERROR|CRITICAL]
                                  Sets the logging level  [env var: LOG_LEVEL;
                                  default: INFO]
  --log_main_only / --no_log_main_only
                                  When set, only logging from the main thread;
                                  useful when there are multiple Accelerate
                                  processes producing too many log entries to
                                  stdout  [env var: LOG_MAIN_ONLY; default:
                                  no_log_main_only]
  --version                       Show the version and exit.
  --help                          Show this message and exit.

Commands:
  eval       Axolotl evaluation tools
  inference  Axolotl inferencing tools
  system     The Axolotl system module
```

Specific command help:

```shell
axolotl system config --help
Usage: axolotl system config [OPTIONS]

  Applies override logic, performs validation, and outputs the derived
  configuration. Useful for developing configuration overrides and unit
  testing.

Options:
  --pretty / --no-pretty          Pretty print the JSON output
  -f, --filter TEXT               Output only JSON keys that match the
                                  provided filter expressions
  --adapter [lora|qlora]          Adapter type  [env var: AXOLOTL_ADAPTER]
  --base_model_config TEXT        Useful when the base_model repo on
                                  HuggingFace hub doesn't include
                                  configuration .json files. When empty,
                                  defaults to config in base_model  [env var:
                                  AXOLOTL_BASE_MODEL_CONFIG]
  --base_model_ignore_patterns TEXT
                                  An ignore pattern if the model repo contains
                                  more than 1 model type (*.pt, etc)  [env
                                  var: AXOLOTL_BASE_MODEL_IGNORE_PATTERNS]
  --base_model TEXT               The huggingface model that contains *.pt,
                                  *.safetensors, or *.bin files or a path to
                                  the model on the disk  [env var:
                                  AXOLOTL_BASE_MODEL]
```

For full documentation see: docs/cli.md

## Multi-GPU inferencing

Batch inferencing takes a list of json-formatted datasets, applies the configured prompt template, executes the inferencing process, and saves the output to ``output_dir``. Inferencing operations distribute prompts across GPUs via [accelerate](https://huggingface.co/docs/accelerate/index).

This feature takes advantage of the new CLI enhancements and can be invoked via:

```shell
accelerate launch --config_file=/work/accelerate/development_gpu_all.yaml -m \
    axolotl -c /defaults.yaml inference batch \
      --dataset=name=DS1,path=/workspace/sample1.json,type=gpteacher
      --dataset=name=DS2,path=/workspace/sample2.json,type=gpteacher
    # Other CLI overrides go here...
```

The output will be in a single JSON-formatted file, ex:

```json
{
  "status": "SUCCESS",
  "run_id": "20230801T130712Z",
  "request_count": 10,
  "response_count": 10,
  "run_time_sec": 12.925,
  "total_tokens_generated": 200,
  "tokens_per_sec": 15.474,
  "seed": 10,
  "response": [
    {
      "prompt_tokens": 101,
      "total_tokens": 121,
      "total_response_tokens": 20,
      "generate_time_sec": 1.416,
      "prompt": "...",
      "responses": [
        {
          "response": "...",
          "tokens": 20
        }
      ]
    },
    {
      "prompt_tokens": 65,
      "total_tokens": 85,
      "total_response_tokens": 20,
      "generate_time_sec": 0.912,
      "prompt": "...",
      "responses": [
        {
          "response": "...",
          "tokens": 20
        }
      ]
    }
  ]
}
```

For full documentation see: docs/inference.md

## Other updates

There are also a few additional updates included in this PR. Without trying to change the codebase too much, these seemed necessary while implementing the eval / inferencing features.

### (yet another) update to logging

Added ``disable_existing_loggers=False`` to logging configuration to enable logging of all modules. I feel like this is what we want and verbosity can be controlled in the LOG_LEVEL itself.

Added Accelerate [logging handler](https://huggingface.co/docs/accelerate/package_reference/logging) to new CLI classes which will help reduce duplicate logs from multiple accelerate processes. Additionally, the logging configuration will automatically set the
``TRANSFORMERS_VERBOSITY`` and ``ACCELERATE_LOG_LEVEL`` to the configured log level to ensure that all logging settings stay in sync.

### New configuration: split_name

Name of the split when loading the dataset from HF. This seemed reasonable since batch_eval / batch_inference use cases we may not to always want to use the "train" split.

### New configuration: truncate_features

Whenever a base JSON dataset is provided for inferencing the axolotl prompt generation logic is invoked. The ``truncate_features`` option will truncate any response data in the json file prior to the prompt being generated. For the inferencing use case it is highly likely this is what the user will want since it doesn't make sense to generate a prompt template for inferencing with the "best" response already present.

### Small refactor deriving default_dataset_prepared_path

In data.load_prepare_datasets() the logic that considered ``cfg.dataset_prepared_path`` when deriving ``default_dataset_prepared_path`` was only applied when ``cfg.max_packed_sequence_len`` was not ``None``. This may have been a normal case however to set the default consistently for all cases, this update moves the logic into ``train()``:

```python
derived_default_dataset_prepared_path = cfg.dataset_prepared_path if cfg.dataset_prepared_path else DEFAULT_DATASET_PREPARED_PATH
LOG.debug("Using derived_default_dataset_prepared_path = %s", derived_default_dataset_prepared_path)
train_dataset, eval_dataset = load_prepare_datasets(
    tokenizer, cfg, derived_default_dataset_prepared_path
)
```

### Added package version

Added an axolotl package version identification string. I considered this part of the CLI updates to support ``axolotl --version`` however it seems like its own thing. Ideally we would dynamically update ``src/axolotl/version.py`` as part of the release process, which is why I put it in its own file.

### CODE REVIEW

**Comment by theobjectivedad** [comment]
Hi @winglian , I see you are [working on adding a perplexity metric](https://github.com/OpenAccess-AI-Collective/axolotl/blob/0f2a16aa33faa5ec73bb783d3e0ac17cb7cd89d9/src/axolotl/utils/trainer.py#L340) in a fb, if you have any time would you mind giving me feedback on this PR, in particular the perplexity calculation? To be honest I know just enough to be dangerous and would love an early review.

File: src/axolotl/cli/batch_eval.py [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+                    loss_reduced = self.accelerator.gather(outputs.loss)
+                    if self.accelerator.is_local_main_process:
+                        total_loss += (
+                            torch.mean(loss_reduced) / self.accelerator.num_processes
+                        )
+
+        # Only main process computes average loss and perplexity
+        if self.accelerator.is_local_main_process:
+            avg_loss = total_loss / len(dataloader)
+            perplexity = torch.exp(avg_loss)
```
Comment (winglian) [inline src/axolotl/cli/batch_eval.py]: this perplexity calculation seems correct according to everything I've seen.

**Comment by eugenepentland** [comment]
Hey theobjectivedad, would you be able to break this into a few smaller PRs? Seems like you've done a lot of good work but its a lot to review all at once. I had been talking to wingland about this PR earlier today.

**Comment by theobjectivedad** [comment]
Good morning @eugenepentland , thank you for looking at my giant PR :D 

I have a few thoughts and see 3 potential splits for this PR: 

1. CLI enhancements, package versioning
2. Batch inferencing, new config options
3. #165 and logging bugfixes

Number 2 is dependent on number 1 however it would be fairly trivial for me to split these into their own PRs. I would however argue that these are 90% new files / new features and merging 1 without 2 wouldn't provide any new capabilities. My recommendation would be to keep these together.

For number 3, this is only ~5%-10% of the overall PR and IMO wouldn't be worth the trouble to split out.

Would it be OK with you if we allocated more time to fully review the PR and/or maybe setup a call and I can walk you & the review team through the codebase to make it less painful?

**Comment by theobjectivedad** [comment]
Good morning @eugenepentland , does this ^^ seem reasonable?

**Comment by theobjectivedad** [comment]
Good morning @eugenepentland and @winglian , this AM I've updated with the changes in ``main``, please let me know the status of this PR when you have a moment. Again, thank you for your time!

**Comment by tmm1** [comment]
hi, I plan to look over this soon as well. sorry for all the merge conflicts!

**Comment by NanoCode012** [comment]
Closing this as we've migrated. Thank you for the PR and starting this off.
